validator/valnode/redis/consumer.go

```diff
@@ -43,46 +44,86 @@ func NewValidationServer(cfg *ValidationServerConfig, spawner validator.Validati
+	readyStreams := make(chan struct{}, len(s.consumers))
 	for moduleRoot, c := range s.consumers {
 		c := c
+		moduleRoot := moduleRoot
 		c.Start(ctx_in)
-		s.StopWaiter.CallIteratively(func(ctx context.Context) time.Duration {
-			req, err := c.Consume(ctx)
-			if err != nil {
-				log.Error("Consuming request", "error", err)
-				return 0
+		// Channel for single consumer, once readiness is indicated in this,
+		// consumer will start consuming iteratively.
+		ready := make(chan struct{}, 1)
+		s.StopWaiter.LaunchThread(func(ctx context.Context) {
+			for {
+				if pubsub.StreamExists(ctx, c.StreamName(), c.RedisClient()) {
+					ready <- struct{}{}
+					readyStreams <- struct{}{}
+					return
+				}
+				time.Sleep(time.Millisecond * 100)
 			}
-			if req == nil {
-				// There's nothing in the queue.
-				return time.Second
-			}
-			valRun := s.spawner.Launch(req.Value, moduleRoot)
-			res, err := valRun.Await(ctx)
-			if err != nil {
-				log.Error("Error validating", "request value", req.Value, "error", err)
-				return 0
-			}
-			if err := c.SetResult(ctx, req.ID, res); err != nil {
-				log.Error("Error setting result for request", "id", req.ID, "result", res, "error", err)
-				return 0
+		})
+		s.StopWaiter.LaunchThread(func(ctx context.Context) {
+			select {
+			case <-ctx.Done():
+				log.Error("Context done", "error", ctx.Err().Error())
+				return
+			case <-ready: // Wait until the stream exists and start consuming iteratively.
 			}
-			return time.Second
+			s.StopWaiter.CallIteratively(func(ctx context.Context) time.Duration {
+				req, err := c.Consume(ctx)
+				if err != nil {
+					log.Error("Consuming request", "error", err)
+					return 0
+				}
+				if req == nil {
+					// There's nothing in the queue.
+					return time.Second
+				}
+				valRun := s.spawner.Launch(req.Value, moduleRoot)
+				res, err := valRun.Await(ctx)
+				if err != nil {
+					log.Error("Error validating", "request value", req.Value, "error", err)
+					return 0
+				}
+				if err := c.SetResult(ctx, req.ID, res); err != nil {
+					log.Error("Error setting result for request", "id", req.ID, "result", res, "error", err)
+					return 0
+				}
+				return time.Second
+			})
 		})
 	}
+
+	for {
+		select {
+		case <-readyStreams:
+			log.Trace("At least one stream is ready")
+			return // Don't block Start if at least one of the stream is ready.
+		case <-time.After(s.streamTimeout):
+			log.Error("Waiting for redis streams timed out")
+		case <-ctx_in.Done():
+			log.Error(("Context expired, failed to start"))
+			return
+		}
+	}
 }
 
 type ValidationServerConfig struct {
 	RedisURL       string                `koanf:"redis-url"`
 	ConsumerConfig pubsub.ConsumerConfig `koanf:"consumer-config"`
 	// Supported wasm module roots.
 	ModuleRoots []string `koanf:"module-roots"`
+	// Timeout on polling for existence of each redis stream.
+	StreamTimeout time.Duration `koanf:"stream-timeout"`
```

Comment:
add a default value (5-15minutes sounds good), and add to ValidationServerConfigAddOptions

Reply:
Done.